server/routes.go

```diff
@@ -1967,88 +1967,153 @@ func (s *Server) ChatHandler(c *gin.Context) {
+						applyStructuredOutputs = true
+						res.Message.Content = ""
+						ch <- res // flush anything remaining in thinking to user. case where there is both thinking and content with no boundary. e.g. token </think> content
+						requestCancel()
+						return
+					}
+					res.Message.Content = remainingContent
 				}
 
-				return
-			}
+				if len(req.Tools) > 0 {
+					toolCalls, content := toolParser.Add(res.Message.Content)
+					if len(content) > 0 {
+						res.Message.Content = content
+					} else if len(toolCalls) > 0 {
+						res.Message.ToolCalls = toolCalls
+						res.Message.Content = ""
+					} else if res.Message.Thinking != "" {
+						// don't return
+					} else {
+						if r.Done {
+							res.Message.Content = toolParser.Content()
+							ch <- res
+						}
+						return
+					}
+				}
 
-			if thinkingState != nil {
-				thinkingContent, remainingContent := thinkingState.AddContent(res.Message.Content)
-				if thinkingContent == "" && remainingContent == "" && !r.Done {
-					// need to accumulate more to decide what to send
+				ch <- res
+			})
+			if err != nil {
+				if applyStructuredOutputs && !structuredOutputsStarted && (errors.Is(err, context.Canceled) || strings.Contains(err.Error(), "context canceled")) && ctx.Err() == nil {
+					// only ignores error if it's a context cancellation due to setting structured outputs
+				} else {
+					slog.Error("chat completion error", "error", err.Error())
+					ch <- gin.H{"error": err.Error()}
 					return
 				}
-				res.Message.Content = remainingContent
-				res.Message.Thinking = thinkingContent
 			}
 
-			if len(req.Tools) > 0 {
-				toolCalls, content := toolParser.Add(res.Message.Content)
-				if len(content) > 0 {
-					res.Message.Content = content
-				} else if len(toolCalls) > 0 {
-					res.Message.ToolCalls = toolCalls
-					res.Message.Content = ""
-				} else if res.Message.Thinking != "" {
-					// don't return
-				} else {
-					if r.Done {
-						res.Message.Content = toolParser.Content()
-						ch <- res
-					}
+			// ignored structured outputs cancellation falls through to here, start a new request with the structured outputs and updated prompt. use the
+			if applyStructuredOutputs && !structuredOutputsStarted {
+				structuredOutputsStarted = true
+				applyStructuredOutputs = false
+				msg := api.Message{
+					Role:     "assistant",
+					Thinking: tb.String(),
+				}
+
+				msgs = append(msgs, msg)
+				prompt, _, err = chatPrompt(c.Request.Context(), m, r.Tokenize, opts, msgs, processedTools, req.Think)
+				if err != nil {
+					slog.Error("chat prompt error applying structured outputs", "error", err)
+					ch <- gin.H{"error": err.Error()}
 					return
 				}
+				// force constraining by terminating thinking header, the parser is already at this state
+				if shouldUseHarmony(m) {
+					prompt += "<|end|><|start|>assistant<|channel|>final<|message|>"
+				}
```

Comment:
This feels pretty hacky given it exposes the internals of the parser. Shouldn't adding the assistant message on line 2102 be enough without adding this extract prompt?

Reply:
Would the prefill features of the renderer (or template for gpt-oss) add this @drifkin ?

Reply:
No - we can't disambiguate between where the user wants the prefill to continue the thinking vs. finish the generation just from the last message. So unless we were to change what the prefill means we have to expose this.

Reply:
OK. Possible to add a TODO?